### ISSUE TITLE
MemorySanitizer reports false positive for uninitialized values in padding

### ISSUE THREAD

**Issue by psamolysov-intel** [llvm/llvm-project#54476]
We have the following code compiled for x86-64:

```c++
#include <cstdint>
#include <cstdio>
#include <vector>

struct  Demo {
  uint32_t Size;
  uint8_t Flag;
};

int main(int argc, char **argv) {
  static constexpr char EncodingTable[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                                          "abcdefghijklmnopqrstuvwxyz"
                                          "0123456789+/";
  std::vector<Demo> V{{{1024, 1}}};
  for (size_t i = 0; i < V.size() * sizeof(Demo); i++) {
    auto idx = reinterpret_cast<const unsigned char *>(V.data())[i];
    printf("%c ", EncodingTable[idx >= sizeof(EncodingTable) ? 0 : idx]);
  }
  printf("\n");

  return 0;
}
```
If the code is compiled with MemorySanitizer support:

```
clang++ -fsanitize=memory -fno-omit-frame-pointer -fno-optimize-sibling-calls -O2 -g -o reproducer reproducer.cpp
```

The sanitizer reports about uninitialized value:

```
./reproducer
==246571==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4ad8f2 in main reproducer.cpp:18:19
    #1 0x7fdd9c63b0b2 in __libc_start_main glibc-2.31/csu/../csu/libc-start.c:308:16
    #2 0x41d2dd in _start (reproducer+0x41d2dd)
```

Line 18:

```c++
    printf("%c ", EncodingTable[idx >= sizeof(EncodingTable) ? 0 : idx]);
```

The problem is the following: structure `Demo` has the last member with the `uint8_t` type, by default the structure is aligned by 8 bytes and 3 bytes were added at the end of the structure as padding. When an instance of the structure is created, nothing is written in the padding bytes, so they aren't initialized and MemorySanitizer reports about this. If the structure is defined as packed, so without any padding:

```c++
struct __attribute__ ((packed)) Demo {
  uint32_t Size;
  uint8_t Flag;
};
```
The sanitizer emits no report, everything works as expected.

This behavior of the sanitizer makes any structure with padding unusable.

One observation, if bytes are just read from the structure, MemorySanitizer emits no report:

```c++
  std::vector<Demo> V{{{1024, 1}}};
  for (size_t i = 0; i < V.size() * sizeof(Demo); i++) {
    auto idx = reinterpret_cast<const unsigned char *>(V.data())[i];
    printf("%d ", idx);
  }
  printf("\n");

  return 0;
```

But if we use an uninitialized value as an index, MemorySanitizer emits the report quoted above.
